include/convex_bodies/ellipsoid.h

```diff
@@ -0,0 +1,307 @@
+        
+        NT D = std::pow(b_q, 2) - 4*a_q*c_q;
+        return std::pair<NT, NT> ( (-b_q + std::sqrt(D))/(2*a_q) , (-b_q - std::sqrt(D))/(2*a_q) );
+    }
+
+
+    std::pair<NT,NT> line_intersect(Point const& r,
+                                    Point const& v,
+                                    const VT &Ar,
+                                    const VT &Av) const
+    {
+        return line_intersect(r, v);
+    }
+
+
+    std::pair<NT,NT> line_intersect(Point const& r,
+                                    Point const& v,
+                                    const VT &Ar,
+                                    const VT &Av,
+                                    NT &lambda_prev) const
+    {
+        return line_intersect(r, v);
+    }
+
+
+    std::pair<NT,int> line_positive_intersect(Point const& r,
+                                              Point const& v) const
+    {
+        return std::pair<NT,NT>(line_intersect(r, v).first, 0);
+    }
+
+
+    std::pair<NT,int> line_positive_intersect(Point const& r,
+                                              Point const& v,
+                                              const VT &Ar,
+                                              const VT &Av) const
+    {
+        return line_positive_intersect(r, v);
+    }
+
+
+    std::pair<NT,int> line_positive_intersect(Point const& r,
+                                              Point const& v,
+                                              const VT &Ar,
+                                              const VT &Av,
+                                              NT &lambda_prev) const
+    {
+        return line_positive_intersect(r, v);
+    }
+
+
+    // Compute the intersection of a coordinate ray
+    std::pair<NT,NT> line_intersect_coord(const Point &r, const unsigned int rand_coord) const {
+        NT a_q = A(rand_coord, rand_coord);
+        NT b_q = 2 * (r - c).getCoefficients().dot(A.col(rand_coord));
+        NT c_q = mat_mult(r - c);
+        
+        NT D = std::pow(b_q, 2) - 4*a_q*c_q;
+        return std::pair<NT, NT> ( (-b_q + std::sqrt(D))/(2*a_q) , (-b_q - std::sqrt(D))/(2*a_q) );
+    }
+
+
+    std::pair<NT,NT> line_intersect_coord(Point const& r,
+                                          unsigned int const& rand_coord,
+                                          const VT &lamdas) const
+    {
+        return line_intersect_coord(r, rand_coord);
+    }
+
+
+    std::pair<NT,NT> line_intersect_coord(Point const& r,
+                                          Point const& r_prev,
+                                          unsigned int const& rand_coord,
+                                          unsigned int const& rand_coord_prev,
+                                          const VT &lamdas) const
+    {
+        return line_intersect_coord(r, rand_coord);
+    }
+
+
+    void compute_reflection (Point& v, Point const& p) const
+    {
+        // normal vector is Ap
+        Point s(A * (p-c).getCoefficients());
```

Comment:
A is symmetric. Could you use Eigen's optimization for the matrix-vector multiplication?

Reply:
added an optimized `vec_mult` function.